### PR TITLE
bpo-34726: Fix handling of hash-based pycs in zipimport

### PR DESCRIPTION
**Pull request by elprans**
Current support for hash-based bytecode files in `zipimport` is rather
sparse, which leads to test failures when the test suite is ran with
the ``SOURCE_DATE_EPOCH`` environment variable set.

This teaches zipimport to handle hash-based pycs properly.

https://bugs.python.org/issue34022

https://bugs.python.org/issue34726

### CODE REVIEW

**Review by ncoghlan — APPROVED** [review]
The new tests look good to me, and the changes required to make them pass seem sensible.

**Comment by ncoghlan** [comment]
Thanks for the PR @elprans!

@serhiy-storchaka @warsaw @brettcannon This looks good to me, but I'd be happier if someone more familiar with the new zipimport code gave it a look before we hit the big green merge button :)

**Comment by serhiy-storchaka** [comment]
I created bpo-34726, but I'm not sure that any checks for pyc files are needed at all. They slow down import, and we can expect that pyc files in a zip archive are up to date with py files. See bpo-23734.

File: Lib/zipimport.py [hunk 1]
```diff
@@ -578,33 +578,53 @@ def _eq_mtime(t1, t2):
...
-        # will use unchecked ones.
+def _unmarshal_code(self, pathname, fullpath, fullname, data):
+    exc_details = {
+        'name': fullname,
+        'path': fullpath,
+    }
+
+    try:
+        flags = _bootstrap_external._classify_pyc(data, fullname, exc_details)
+    except ImportError as e:
```
Comment (serhiy-storchaka) [inline Lib/zipimport.py]:
```suggestion
    except ImportError:
```

File: Lib/zipimport.py [hunk 2]
```diff
@@ -651,9 +671,25 @@ def _get_mtime_of_source(self, path):
...
+# contents of the matching .py file, or None if no source
+# is available.
+def _get_pyc_source(self, path):
+    try:
+        # strip 'c' or 'o' from *.py[co]
+        assert path[-1:] in ('c', 'o')
+        path = path[:-1]
+        toc_entry = self._files[path]
+        return _get_data(self.archive, toc_entry)
     except (KeyError, IndexError, TypeError):
```
Comment (serhiy-storchaka) [inline Lib/zipimport.py]: In what circumstances IndexError and TypeError are raised?

File: Misc/NEWS.d/next/Core and Builtins/2018-11-04-18-13-40.bpo-34022.U3btVj.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix handling of hash-based bytecode files in :mod:`zipimport`.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2018-11-04-18-13-40.bpo-34022.U3btVj.rst:1]: Add "Patch by ..."

File: Lib/zipimport.py (hunk 2 above)
Comment (elprans) [reply Lib/zipimport.py]: I copied this from `_get_mtime_of_source()` without really thinking.  I'm guessing only `KeyError` is relevant in this case (raised by `self._files[path]`).

**Comment by serhiy-storchaka** [comment]
You need to rebuild frozen modules. Run `make regen-importlib`.

**Comment by elprans** [comment]
Done

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by vstinner** [comment]
test_zipfile_compiled_checked_hash() fails on Python 3.7: we should backport this bugfix to 3.7, no?

**Comment by vstinner** [comment]
Ah, the change cannot be cherry-picked automatically. Maybe because of the generated file.

**Comment by elprans** [comment]
@vstinner I'll make a backport PR.

**Comment by elprans** [comment]
It's a bit more difficult than that.  In 3.7 `zipimport` is in C.

**Comment by serhiy-storchaka** [comment]
It is easier to skip the test on 3.7.

**Pull request merged**